### ISSUE TITLE
如何在线程函数内创建窗口?

### ISSUE THREAD

**Issue by u0yiji1** [netease-im/NIM_Duilib_Framework#152]
我想在线程函数内创建一个新的窗口出来 但是创建的窗口不显示

void showfrom()
{
    TestForm* Test = new TestForm();
    if (Test->Create(NULL, _T("Test"), UI_WNDSTYLE_DIALOG, 0))
    {
        Test->CenterWindow();
        Test->ShowWindow(true);
    }
}

bool MainForm::OnClicked(ui::EventArgs * msg)
{
    if (msg->pSender->GetName() == L"button1")
    {
        showfrom();//这样会显示
        std::thread(showfrom).detach();//这样不会显示
    }
}

**Comment by nmgwddj** [comment]
你的线程还是可能还没有执行就已经出了作用域，可以试试作为成员变量。

另外，强烈建议您使用 Base 的线程管理相关接口来做多线程的操作，将 UI 相关操作全部投递到 UI 线程中。

## 参考

投递任务到 UI 线程：https://github.com/netease-im/NIM_Duilib_Framework/blob/master/examples/controls/controls_form.cpp#L106
创建一个杂事线程：https://github.com/netease-im/NIM_Duilib_Framework/blob/master/examples/controls/main.cpp#L35

**Comment by u0yiji1** [comment]
多谢指点

**Comment by u0yiji1** [comment]
@nmgwddj   使用 PostTask(kThreadUI 会卡界面 

void MainForm::Test()
{
	std::this_thread::sleep_for(std::chrono::milliseconds(2500));
	TestForm* Test = new TestForm();
	if (Test->Create(NULL, _T("Test"), UI_WNDSTYLE_DIALOG, 0))
	{
		Test->CenterWindow();
		Test->ShowWindow(true);
	}
}

nbase::ThreadManager::PostTask(kThreadUI, nbase::Bind(&MainForm::Test));

使用杂事线程的情况下虽然不会卡界面 依然是不能显示窗口 

class MiscThread : public nbase::FrameworkThread
{
public:
	MiscThread(int thread_id, const char *name)
		: FrameworkThread(name)
		, thread_id_(thread_id) {}

	~MiscThread(void) {}
private:
	virtual void Init() override;
	virtual void Cleanup() override;
private:
	int thread_id_;
};

void MiscThread::Init()
{
	nbase::ThreadManager::RegisterThread(thread_id_);
}

void MiscThread::Cleanup()
{
	nbase::ThreadManager::UnregisterThread();
}

nbase::ThreadManager::PostTask(kThreadGlobalMisc, nbase::Bind(&MainForm::Test));

**Comment by xuhuajie-NetEase** [comment]
那是因为开窗口的线程没有制造自己的消息循环。
当然也要加上自己的窗口过程函数，是可以实现多线程多窗口的。
但要注意小心崩溃，原则上不允许多线程操作同一窗口
MiscThread窗口的窗口的UI操作，都必须nbase::ThreadManager::PostTask(kThreadGlobalMisc,
试试在create后加上
BOOL bRet;

while( (bRet = GetMessage( &msg, hWnd, 0, 0 )) != 0)
{ 
    if (bRet == -1)
    {
        // handle the error and possibly exit
    }
    else
    {
        TranslateMessage(&msg); 
        DispatchMessage(&msg); 
    }
}

**Comment by xuhuajie-NetEase** [comment]
在 Win32 中，每一个线程有它自己专属的消息队列。
在create后面加上
BOOL bRet;

while( (bRet = GetMessage( &msg, hWnd, 0, 0 )) != 0)
{ 
    if (bRet == -1)
    {
        // handle the error and possibly exit
    }
    else
    {
        TranslateMessage(&msg); 
        DispatchMessage(&msg); 
    }
}
试试

**Comment by nmgwddj** [comment]
在 Controls 工程中添加了一个[示例](https://github.com/netease-im/NIM_Duilib_Framework/commit/0702ea2cf442c061dff78094bb476f99c2b3687c)，演示了如何在杂事线程执行任务完成后创建其他窗口的代码。具体步骤如下：

 - 投递一个杂事任务来加载一些耗时任务：https://github.com/netease-im/NIM_Duilib_Framework/blob/development/examples/controls/controls_form.cpp#L88
 - 在耗时任务结束后投递单独的任务到 UI 线程去更新 UI 资源，保证 UI 相关操作都在一个线程内而不会引起冲突：https://github.com/netease-im/NIM_Duilib_Framework/blob/development/examples/controls/controls_form.cpp#L168
 - 在投递到 UI 线程热任务函数中创建了另外一个窗口：https://github.com/netease-im/NIM_Duilib_Framework/blob/development/examples/controls/controls_form.cpp#L185

其中创建新窗口的代码使用了我们封装的接口，如无特殊需求这样使用非常方便，你同样可以参考另外一些示例代码，查看如何自定义创建一个窗口：https://github.com/netease-im/NIM_Duilib_Framework/blob/development/examples/controls/controls_form.cpp#L123

**Comment by u0yiji1** [comment]
多谢
